### ISSUE TITLE
Printing some R objects with custom `print` method fails with `unused argument (useS4 = FALSE)`

### ISSUE THREAD

**Issue by krassowski** [rpy2/rpy2#908]
**Describe the issue or bug**

Using `methods::show` as a default proxy for R's print leads custom print representations to error out, for example in easystats' parameters: https://github.com/easystats/parameters/issues/747. According to https://github.com/easystats/parameters/issues/747#issuecomment-1190270611, the preferred solution would be to deal with it downstream in rpy2 as quote:

> `show()` is a function for S4 objects (that's what the _methods_ package is--the home of the S4 system). It isn't intended as a drop in replacement for `print()` and shouldn't be used as such. S4 is a largely dead part of R, and its idiosyncrasies aren't something that is or probably should be widely supported.
> 
> You shouldn't expect it to work with non-S4 objects or use it in general. If an interface is using `show()` and that causes problems, the inference should fix that by using R's actual function for displaying output--`print()`.

**To Reproduce**

1. Install `parameters` from `easystats`: https://github.com/easystats/parameters
2. Run the following code:
    ```python
    from rpy2.robjects.packages import importr
    from rpy2.robjects import r
    
    stats = importr('stats')
    parameters = importr('parameters')
    iris = r['iris']
    
    model = stats.lm(stats.as_formula('Sepal.Width ~ Petal.Length * Species + Petal.Width'), data = iris)
    result = parameters.model_parameters(model)
    print(result)
    ```
3. See error: `R[write to console]: Error in insight::export_table(formatted_table, format = "text", caption = table_caption,  : unused argument (useS4 = FALSE)`

**Expected behavior**
Displaying R objects work consistently.

Possible solutions:
- use `base::print` instead of `methods::show` and catch the stdout
- use `repr::repr` from [repr](https://github.com/IRkernel/repr) or `knitr::knit_print` from [knitr](https://github.com/yihui/knitr) as previously discussed in https://github.com/rpy2/rpy2/issues/646 (I think repr is lightweight enough to be an optional but recommended dependency)
- wrapping `methods::show` call in try-except and if we match `unused argument (useS4 = FALSE)`, then use base `print`.

**Error**

```python
R[write to console]: Error in insight::export_table(formatted_table, format = "text", caption = table_caption,  : 
  unused argument (useS4 = FALSE)
--------------------------------------------------------------------------
RRuntimeError                             Traceback (most recent call last)
/tmp/ipykernel_380637/2276435863.py in <module>
      1 model = stats.lm(stats.as_formula('Sepal.Width ~ Petal.Length * Species + Petal.Width'), data = iris)
      2 result = parameters.model_parameters(model)
----> 3 print(result)

~/rpy2/rpy2/robjects/robject.py in __str__(self)
    110               .callbacks.obj_in_module(rpy2.rinterface_lib.callbacks,
    111                                        'consolewrite_print', s.append)):
--> 112             self.__show(self)
    113         s = str.join('', s)
    114         return s

~/rpy2/rpy2/rinterface_lib/conversion.py in _(*args, **kwargs)
     43 def _cdata_res_to_rinterface(function):
     44     def _(*args, **kwargs):
---> 45         cdata = function(*args, **kwargs)
     46         # TODO: test cdata is of the expected CType
     47         return _cdata_to_rinterface(cdata)

~/rpy2/rpy2/rinterface.py in __call__(self, *args, **kwargs)
    811             )
    812             if error_occured[0]:
--> 813                 raise embedded.RRuntimeError(_rinterface._geterrmessage())
    814         return res
    815 

RRuntimeError: Error in insight::export_table(formatted_table, format = "text", caption = table_caption,  : 
  unused argument (useS4 = FALSE)
```

**Additional context**

This was previously discussed in https://github.com/rpy2/rpy2/issues/763, but there the upstream maintainers agreed to a fix.

In rpy2, the relevant code is in:

https://github.com/rpy2/rpy2/blob/796843f34a54634cd286ca782a8b232b2ee8d53c/rpy2/robjects/robject.py#L92

https://github.com/rpy2/rpy2/blob/796843f34a54634cd286ca782a8b232b2ee8d53c/rpy2/robjects/robject.py#L106-L114

**Comment by lgautier** [comment]
> **Describe the issue or bug**
> 
> Using `methods::show` as a default proxy for R's print leads custom print representations to error out, for example in easystats' parameters: [easystats/parameters#747](https://github.com/easystats/parameters/issues/747). According to [easystats/parameters#747 (comment)](https://github.com/easystats/parameters/issues/747#issuecomment-1190270611), the preferred solution would be to deal with it downstream in rpy2 as quote:
> 
> > `show()` is a function for S4 objects (that's what the _methods_ package is--the home of the S4 system). It isn't intended as a drop in replacement for `print()` and shouldn't be used as such. S4 is a largely dead part of R, and its idiosyncrasies aren't something that is or probably should be widely supported.
> > You shouldn't expect it to work with non-S4 objects or use it in general. If an interface is using `show()` and that causes problems, the inference should fix that by using R's actual function for displaying output--`print()`.

`S4` was once the up-and-coming to do OOP in R. I have not followed its development in a while and I understand that it can be disliked but news about its death might be greatly exaggerated. The error triggered by rpy2 should be fixed though.

> **To Reproduce**
> 
>     1. Install `parameters` from `easystats`: https://github.com/easystats/parameters
> 
>     2. Run the following code:
>        ```python
>        from rpy2.robjects.packages import importr
>        from rpy2.robjects import r
>        
>        stats = importr('stats')
>        parameters = importr('parameters')
>        iris = r['iris']
>        
>        model = stats.lm(stats.as_formula('Sepal.Width ~ Petal.Length * Species + Petal.Width'), data = iris)
>        result = parameters.model_parameters(model)
>        print(result)
>        ```
> 
>     3. See error: `R[write to console]: Error in insight::export_table(formatted_table, format = "text", caption = table_caption,  : unused argument (useS4 = FALSE)`
> 
> 
> **Expected behavior** Displaying R objects work consistently.
> 
> Possible solutions:
> 
>     * use `base::print` instead of `methods::show` and catch the stdout

I remember using `print` at a much earlier point in time, and moving to `show` because it was better. I don't remember more details though, therefore to which extent it was critically better. I should check whether this still holds.

>     * use `repr::repr` from [repr](https://github.com/IRkernel/repr) or `knitr::knit_print` from [knitr](https://github.com/yihui/knitr) as previously discussed in [Allow rich/custom display of result from R cell (R magic) #646](https://github.com/rpy2/rpy2/issues/646) (I think repr is lightweight enough to be an optional but recommended dependency)

My first reaction is to avoid an additional dependency.

>     * wrapping `methods::show` call in try-except and if we match `unused argument (useS4 = FALSE)`, then use base `print`.

Actually I think that the error should likely be fixed upstream (in R, or R package). `useS4` does not seem to originate from `rpy2` and `show` should work with *any* R object. The doc for `show()` is:

```
Usage:

     show(object)

Arguments:

  object: Any R object
```

**Comment by lgautier** [comment]
I am adding a fallback to R's `print()` with a warning message. This is still likely an issue upstream. I am lacking the bandwidth to chase this. Great if someone can.

**Comment by lgautier** [comment]
Fixed with PR #950.
